File: ML/Ml.jl/Decision-Tree.jl
using DataFrames, CSV;
using Serialization;

function calcEnt(datasets)
    num_samples = size(datasets,1)
    label_counts = Dict()
    for label in datasets[end]
        if label in keys(label_counts)
            label_counts[label] += 1
        else
            label_counts[label] = 1
        end
    end
    # $ Ent(D) = - ∑_{p ∈ P} p log_2(p) $
    return - sum([p * log2(p) for p in values(label_counts) ./ num_samples ])
end

function calcGain(datasets,class,ent = calcEnt(datasets))
    num_samples = size(datasets,1)
    groups = groupby(datasets, class)
    entries = [(size(subsets,1),calcEnt(subsets)) for subsets in groups]
    # $Gain(D,a) = Ent(D) - ∑_{v ∈ V} |D^v| / |D| Ent(D^v)$
    return ent - sum([group_size / num_samples * grout_ent for (group_size,grout_ent) in entries])
end

function calcGainRatio(datasets,class,ent = calcEnt(datasets))
    num_samples = size(datasets,1)
    groups = groupby(datasets, class)
    entries = [(size(subsets,1),calcEnt(subsets)) for subsets in groups]
    gain =  ent - sum([group_size / num_samples * grout_ent for (group_size,grout_ent) in entries])
    # $IV = - ∑^V_{v=1} |D^v| / |D| log2(|D^v| / |D|)$
    IV = - sum([group_size / num_samples * log2(group_size / num_samples) for (group_size,grout_ent) in entries])
    # $Gain(D,a) = Gain(D,a) / IV$
    return gain / IV
end

function maxLabel(datasets)
    return by(datasets, :好瓜, df -> DataFrame(N = size(df, 1)))
end

function createTree(datasets, classes = names(datasets))
    groups = groupby(datasets,classes[end])
    if length(groups) == 1          # 取出分类标签（好瓜/坏瓜），如果只有一种，不需要继续分割
        return groups[end][end][end]
    end
    if length(classes) == 1         # 没有别的类别可以用来分割，取概率最大的分类标签
        max_group = groups[1]
        for group in groups[2:end-1]
            if length(group) > length(max_group)
                max_group = group
            end
        end
        return max_group[end][end]
    end

    best_class = classes[1]         # 取出最好的分类类别，即信息增益最大的类别
    best_gain = calcGain(datasets,best_class)
    for class in classes[1:end-1]
        curr_gain = calcGain(datasets,class)
        if curr_gain > best_gain
            best_class = class
            best_gain = curr_gain
        end
    end

    new_classes = copy(classes)     # 递归调用，下一个分支需要用到的分类标签
    filter!(x->x!=best_class, new_classes)   # 由于julia传引用，需要拷贝一份
    best_class_label = Set(datasets[best_class])
    tree = Dict(best_class => Dict())

    # 遍历该类别标准下所有子分类，递归求解
    for group in groupby(datasets,best_class)
        tree[best_class][group[best_class][1]] = createTree(group,new_classes)
    end

    return tree
end

datasets = CSV.read("Work/ML/2.0.csv");
tree = createTree(datasets)

print(tree)

describe(datasets)
by(datasets, :好瓜, df -> DataFrame(N = size(df, 1)))

stream = open("Work/ML/tree.data","w")
serialize(stream,tree)
close(stream)

stream = open("Work/ML/tree.data","r")
read_tree = deserialize(stream)
close(stream)

File: ML/Ml.py/decision_tree.py
from math import log2
import json

import pandas
from pandas import DataFrame


def probe(func):
    name = func.__name__.strip("calc")

    def wrapper(*args, **kwargs):
        result = func(*args, **kwargs)
        if len(args) >= 2:
            print(f"${name}(D, \\text{'{'}{args[1]}{'}'}) = {result:.3f}$")
        else:
            print(name, result)
        return result

    return wrapper


def calcEnt(datasets: DataFrame, class_=None):
    total_num = len(datasets)
    label_counts = datasets.groupby(datasets.columns[-1]).size()
    # $ Ent(D) = - \sum_{p \in P} p \log_2(p) $
    ent = -sum([(num / total_num) * log2(num / total_num) for num in label_counts])
    return ent


@probe
def calcGain(datasets: DataFrame, class_, ent=None):
    if ent is None:
        ent = calcEnt(datasets)
    total_size = len(datasets)
    groups = datasets.groupby(class_).apply(lambda x: (len(x), calcEnt(x)))

    # $Gain(D,a) = Ent(D) - \sum_{v \in V} \frac{ |D^v| }{ |D| Ent(D^v) }$
    return ent - sum(
        [(group_size / total_size) * group_ent for group_size, group_ent in groups]
    )


@probe
def calcGainRatio(datasets: DataFrame, class_, ent=None):
    if ent is None:
        ent = calcEnt(datasets)
    total_size = len(datasets)
    groups = datasets.groupby(class_).apply(lambda x: (len(x), calcEnt(x)))

    # $Gain(D,a) = Ent(D) - \sum_{v \in V} \frac{ |D^v| }{ |D| Ent(D^v) }$
    gain = ent - sum(
        [group_size / total_size * group_ent for group_size, group_ent in groups]
    )

    # $IV = - \sum^V_{v=1} \frac{ |D^v| }{ |D| } log2( \frac{ |D^v| }{ |D| })$
    IV = -sum(
        [
            group_size / total_size * log2(group_size / total_size)
            for (group_size, grout_ent) in groups
        ]
    )
    # $GainRatio(D,a) = \frac{ Gain(D,a) }{ IV }$
    return gain / IV


def createTree(datasets: DataFrame, gain=calcGain, classes=None):
    if classes is None:
        classes = datasets.columns[:-1]

    groups = datasets.groupby("好瓜")

    if len(groups) == 1:  # 取出分类标签（好瓜/坏瓜），如果只有一种，不需要继续分割
        keys = list(groups.groups.keys())
        return keys[0]

    if len(classes) == 1:  # 没有别的类别可以用来分割，取概率最大的分类标签
        return groups.size().idxmax()

    class_gains = {class_: gain(datasets, class_) for class_ in classes}
    best_class, best_gains = max(class_gains.items(), key=lambda x: x[1])

    print(f"=========== Best: {best_class}:{best_gains} ===========")

    tree = {best_class: {}}
    search_classes = [key for key in classes if key != best_class]
    for sub_type, group in datasets.groupby(best_class):
        sub_tree = createTree(group, gain=gain, classes=search_classes)
        tree[best_class][sub_type] = sub_tree

    return tree


def main():
    dataset = pandas.read_csv("data/2.0.csv")

    tree1 = createTree(dataset)
    print(json.dumps(tree1, ensure_ascii=False, indent=2))
    tree2 = createTree(dataset, gain=calcGainRatio)
    print(json.dumps(tree2, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
